Brootux hornet deb packaging

Brootux opened the pull request:
Part of #60 

Rudimentary changes to build deb packages via goreleaser.

WIP and untested, installation should be possible but i could not test fully because systemd does not work in docker and i have currently no dedicated machine to test this.

Temporarily added script "manual_build.sh" to allow local building of deb packages in git-dirty environment and without uploading results. This script should be removed before merge.

RFC

Brootux commented on deb_files/download_snapshot.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+if [ ! -f latest-export.gz.bin ]; then
+	wget https://dbfiles.iota.org/mainnet/hornet/latest-export.gz.bin
+fi
```
Is this how we should handle this? Or should we download the latest-export on every service start?

muXxer replied:
I was thinking about a "run.sh" script, which asks the user if he wants to delete the mainnetdb. If so, it checks automatically if the latest snapshot is still the latest. I have some codesnippets for you if you want to do that? :)

`run.sh`
```sh
#!/bin/bash
service hornet stop

echo "Do you want to delete the old database?"
select yn1 in "Yes" "No"; do
    case $yn1 in
        Yes ) rm -Rf mainnetdb/; ./download_snapshot.sh; break;;
         No ) break;;
    esac
done

service hornet start
journalctl -u hornet -f
```

`download_snapshot.sh`
```sh
#!/bin/bash

finalurl() { curl --silent --location --head --output /dev/null --write-out '%{url_effective}' -- "$@"; }
checkIfModified() { curl --silent --output /dev/null --write-out '%{http_code}' -I --header 'If-Modified-Since: Tue, 17 Dec 2019 11:56:38 GMT' -- "$@"; }

lastest_file=`finalurl https://dbfiles.iota.org/mainnet/hornet/latest-export.gz.bin`
status_code=`checkIfModified ${lastest_file}`
if [[ "$status_code" -ne 304 ]] ; then
    echo "Snapshot outdated, downloading snapshot..."
    rm latest-export.gz.bin
    wget ${lastest_file}
else
    echo "Snapshot up to date"
fi
```

Brootux replied:
When we use systemd as init-system, i think there will be no interactive part, it only starts a service and prints stdout into "syslog" where you can see it i.e. via `journalctl -fu hornet`

Brootux replied:
Nice snippet for outdated snapshot, i will put this in at least

Brootux replied:
Meanwhile there is "wget -N" which activates timestamping and only downloads if remote file is newer.

Brootux commented on manual_build.sh:
```diff
@@ -0,0 +1,9 @@
+
+GORELEASER_IMAGE=iotmod/goreleaser-cgo-cross-compiler:1.13.5
+REPO_PATH="/build"
+PWD="$(pwd)"
```
only usable if executed inside the root-folder of the repository

nuriel77 replied:
Isn't `$PWD` already set to the same value as `$(pwd)`?
```
root@ubuntu-4gb-fsn1-1:~# pwd
/root
root@ubuntu-4gb-fsn1-1:~# echo $PWD
/root
```

muXxer commented on deb_files/hornet.service:
```diff
@@ -0,0 +1,11 @@
+WorkingDirectory=/var/run/hornet
+Restart=always
+ExecStartPre=/usr/bin/hornet_download_snapshot.sh
+ExecStart=/usr/bin/hornet -c /etc/hornet.conf
```
isn't it `/etc/gohornet.conf` ?

svenger87 replied:
Should be /etc/gohornet/gohornet.conf 
Also Timeoutsec=Infinity should be set to let hornet gracefully shutdown.

Brootux reviewed (commented):
Different questions regarding the deb-package:

- How should we handle the `/etc/hornet/hornet.conf` file? it probably gets changed with different default values in the future and needs to replaced or extended. We could place the default conf as `hornet.conf.example` and the user must copy it to `hornet.conf` with adjusted neighbors etc
- I will build it now so that after initial install, the user has to put the service into autostart by hand and also has to start it initially by hand (`systemctl enable hornet` and `systemctl start hornet`). So a user can adjust the hornet-conf first before starting the service. How should we handle the "upgrade" path?
- I would write a very short "getting started" guide, where should i place this? in the readme.md?
- Should hornet run as root or as a special user?

Brootux reviewed (commented):
Cleaned up and tested on AMD64, feel free to test it on your own via `manual_build.sh`

rundirectory is `/var/lib/hornet` where also latest-export gets stored. On `apt purge hornet` this is currently the only directory which does not get deleted. If desired we could add cleanup to postrm-purge

nuriel77 commented on deb_files/hornet.service:
```diff
@@ -0,0 +1,12 @@
+ExecStart=/bin/bash -c "/usr/bin/hornet_download_snapshot.sh && /usr/bin/hornet -c /etc/hornet/hornet.conf"
+
+[Install]
+WantedBy=multi-user.target
```
I had some ideas for the service file.
Use environment file makes it easy to configure some values (no need to issue `systemctl daemon-reload` on each change, as you would to the service file)
```
[Unit]
Description=Hornet
Wants=network-online.target
After=network-online.target

[Service]
EnvironmentFile=-/etc/default/hornet
# Use a variable from the environment file, e.g:
TimeoutStopSec=${TIMEOUT_STOP_SEC}
```
Maybe instead of:
```
ExecStart=/bin/bash -c "/usr/bin/hornet_download_snapshot.sh && /usr/bin/hornet -c /etc/hornet/hornet.conf"`
```
do:
```
ExecStartPre=-/bin/bash /usr/bin/hornet_download_snapshot.sh
ExecStart=/path/to/bin/hornet $OPTIONS # options can be configured in the environment file, should the user want to add
```

Perhaps add (and pre-create the user group so not to run as root):
```
User=hornet
Group=hornet
WorkingDirectory=/path/to/hornet/workdir
```

Add security, logging, optional open files limits:
```
LimitNOFILE=4096
StandardOutput=syslog
StandardError=syslog
SyslogIdentifier=hornet

PrivateDevices=yes
PrivateTmp=yes
ProtectSystem=full
ProtectHome=yes
```

nuriel77 commented on deb_files/postinst:
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+
+set -e
```
You've used `|| true` on most commands, then why `set -e`?

nuriel77 commented on deb_files/postinst:
```diff
@@ -0,0 +1,20 @@
+        systemctl --system daemon-reload >/dev/null || true
+        if [ -n "$2" ]; then
+            deb-systemd-invoke restart "${SYSTEMD_SERVICE}" >/dev/null || true
+	fi
```
indentation +2 spaces

nuriel77 commented on deb_files/postrm:
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+set -e
```
Same, why use `set -e` if defaulting all commands to `|| true` if they fail?
Would make sense to remove `|| true` and let `set -e` do its job, or add proper error handling for commands.

nuriel77 commented on manual_build.sh:
```diff
@@ -0,0 +1,9 @@
+REPO_PATH="/build"
+PWD="$(pwd)"
+
+mkdir -p ${PWD}
```
Always good to use double quotes on variables in bash unless explicitly not required (e.g. multiple cli options provided variable)

Brootux committed:
Added generation of deb-packages

Brootux committed:
Added script to manually build snapshot via goreleaser

Brootux commented:
While testing the package i found out that every time my node was synced and then restarted, i got these "please reset the database" error, is this currently normal?

IoTMOD commented:
This should only happen if HORNET has received a kill signal (by systemd etc.) because the shutdown took too long. The shutdown may take a few minutes depending on the hardware.
Or there was a OOM while shutting down.

Brootux commented:
- Rebased to master
- Found the problem regarding the service restart
- Adjusted for the new confdir/conf-string/neighbors-string parameter interface
- Fixed bugs regarding the helper-script

Removing WIP since i would consider current state as ready

IoTMOD commented on .goreleaser.yml:
```diff
@@ -130,6 +130,35 @@ archives:
+-
+  id: nfpm
+  package_name: hornet
+  file_name_template: "{{ .ProjectName }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}"
```
We should use the dpkg name convention:
https://manpages.debian.org/buster/dpkg-dev/dpkg-name.1.en.html

```suggestion
  file_name_template: "{{ .ProjectName }}_{{ .Version }}_{{ .Arch }}"
```

IoTMOD commented on scripts/build_goreleaser_snapshot.sh:
```diff
@@ -0,0 +1,20 @@
+REPO_PATH="/build"
+
+docker pull "${GORELEASER_IMAGE}"
+docker run --rm --privileged -v "${PWD}":"${REPO_PATH}" -w "${REPO_PATH}" "${GORELEASER_IMAGE}" goreleaser --rm-dist --snapshot
```
Please add a `--skip-publish` (e.g. to not publish a docker image by mistake, if you are logged in into the docker hub)
```suggestion
docker run --rm --privileged -v "${PWD}":"${REPO_PATH}" -w "${REPO_PATH}" "${GORELEASER_IMAGE}" goreleaser --rm-dist --snapshot --skip-publish
```

IoTMOD reviewed (commented):
Hi, 
that's looking good :+1:
In addition to the two suggestions, please change the PR target to the `develop` branch

Brootux commented:
@IoTMOD fixed all your recommendations and targeting `develop` branch now

IoTMOD committed:
format .goreleaser file

IoTMOD committed:
Add short note on how to start HORNET the first time

IoTMOD committed:
Remove /etc/hornet dir with postrm

IoTMOD committed:
Merge branch 'develop' into brootux-hornet-deb-packaging

IoTMOD commented:
Thank you very much!
